consumer/consumererror/README.md

```diff
@@ -0,0 +1,264 @@
+of partial success), it should include this information with a count of the
+failed records.
+
+**Usage:**
+
+```go
+consumererror.WithPartial(failedRecords)
+```
+
+### Indicating error codes from network transports
+
+If the failure occurred due to a network transaction, the exporter should record
+the status code of the message received from the backend. This information can
+be then relayed to the receiver caller if necessary. Note that when the upstream
+component reads a code, it will read a code for its desired transport, and the
+code may be translated depending whether the input and output transports are
+different. For example, a gRPC exporter may record a gRPC status. If a gRPC
+receiver reads this status, it will be exactly the provided status. If an HTTP
+receiver reads the status, it wants an HTTP status, and the gRPC status will be
+converted to an equivalent HTTP code.
+
+**Usage:**
+
+```go
+consumererror.WithGRPCStatus(codes.InvalidArgument)
+consumererror.WithHTTPStatus(http.StatusTooManyRequests)
+```
+
+### Including custom data
+
+Custom data can be included as well for any additional information that needs to
+be propagated back up the pipeline. It is up to the consuming component if or
+how this data will be used.
+
+**Usage:**
+
+```go
+consumererror.WithMetadata(MyMetadataStuct{})
+```
+
+To keep error analysis simple when looking at an error upstream in a pipeline,
+the component closest to the source of an error or set of errors should make a
+decision about the nature of the error. The following are a few places where
+special considerations may need to be made.
+
+## Using errors
+
+### Fanouts
+
+When a fanout receives multiple errors, it will combine them with
+`(consumererror.Error).Combine(errs...)` and pass them upstream. The upstream
+component can then later pull all errors out for analysis.
+
+### Retrieving errors
+
+When a receiver gets a response that includes an error, it can get the data out
+by doing something similar to the following. Note that this uses the `ErrorData`
+type, which is for reading error data, as opposed to the `Error` type, which is
+for recording errors.
+
+```go
+cerr := consumererror.Error{}
+var errData []consumerError.ErrorData 
+
+if errors.As(err, &cerr) {
+  errData := cerr.Data()
+
+  for _, data := range errData {
+    data.HTTPStatus()
+    data.Retryable()
+    data.Partial()
+  }
+}
+```
+
+### Error data
+
+Obtaining data from an error can be done using an interface that looks something like this:
+
+```go
+type ErrorData interface {
+  // Returns the underlying error
+  Error() error
+
+  // Returns nil if a status is not available.
+  HTTPStatus() *int
+
+  // Returns nil if a status is not available.
+  GRPCStatus() *status.Status
+
+  // Returns nil if the error contains no retry information.
+  Retryable() *Retryable
+
+  // Returns a count of failed records or nil if no partial success information was recorded. 
+  Partial() *int
```

Comment:
Shouldn't we have a `Partial` struct similar to `Retryable` so we can extend it if needed?

Reply:
Good with me, done.